bin/monitor

```diff
@@ -2,62 +2,96 @@
 # frozen_string_literal: true
 
 require_relative "../loader"
+clover_freeze
 
 MONITORABLE_RESOURCE_TYPES = [PostgresServer, MinioServer, GithubRunner]
+resources = {}
+mutex = Mutex.new
 
-sessions = {}
-ssh_threads = {}
-pulse_threads = {}
+producer = Thread.new do
+  loop do
+    mutex.synchronize do
+      MONITORABLE_RESOURCE_TYPES.flat_map { _1.all }.each do |r|
+        resources[r.id] ||= {
+          model: r,
+          session: r.reload.init_health_monitor_session,
+          ssh_thread: nil,
+          pulse: {}
+        }
+      rescue Sequel::NoExistingObject
+      end
+    end
+    sleep 5 * 60
+  end
+rescue => ex
+  Clog.emit("Resource loading is failed.") { {health_monitor_resource_load_failure: {exception: Util.exception_to_hash(ex)}} }
 
-clover_freeze
+  # Since we are in a thread, raised exceptions will not be sent to the main
+  # thread, but the thread we are in will exit, causing not to be able to
+  # monitor new resources. We are exiting the whole process to restart it.
+  # Alternatively, we could just ignore the exception and retry, this keeps
+  # the same behavior as the respirate.
+  ThreadPrinter.run
+  Kernel.exit!
+end
 
-loop do
-  resources = MONITORABLE_RESOURCE_TYPES.flat_map { _1.all }
+consumer = Thread.new do
+  loop do
+    Clog.emit("Health checks couldn't be completed.") { {health_monitor_overload: {incomplete_health_check_count: (Thread.list.count - 3) / 2}} } if Thread.list.count > 3
+    mutex.synchronize do
+      keys_to_delete = []
+      resources.each do |r_id, r|
+        while Thread.list.count + 2 > Config.max_monitor_threads
+          sleep 1
+        end
 
-  resources.each do |r|
-    break if Thread.list.count + 2 > Config.max_monitor_threads
-    ssh_threads[r.id] ||= Thread.new do
-      loop do
-        sessions[r.id] = r.init_health_monitor_session
+        next if r[:ssh_thread]&.alive?
```

Comment:
- nil happens in the first iteration, because ssh_thread will be initialized in next line.
- !alive would hopefully happen in all iterations.

Here is the overall logic;

- Start ssh thread (at line 50)
- Start pulse thread (at line 68)
- When one pulse check is completed, kill ssh thread and exit (notice the lack of loop in pulse thread)

That means after each pulse check, we clean up the ssh thread and pulse thread. The reason is not holding the threads while not doing pulse check (which was the weakness of the previous version).

However, since these threads are non blocking, after they are started, outer loop would continue to run. After we iterate over all resources (the loop in line 43) , we sleep for 5 seconds and start another resource iteration, because there is an outermost consumer loop (line 39). It is possible that at the time of second iteration, previous pulse check is not completed (it might take a long time due to various reasons). If this is the case, we should not create new ssh and pulse threads and just wait previous pulse check to complete.

I'll also add a comment about this, but waiting for your ack first.

Reply:
Oof, thanks for that note about `.kill`. Alas, that is an extremely sensitive thing to be doing, because it tends to be implemented with injected exceptions, which is related to the problem of making signal-safe programs. It's related to why I go through all that trouble to `Kernel.exit!` a process rather than using or continuing operation after a Ruby Timeout.

http://headius.blogspot.com/2008/02/rubys-threadraise-threadkill-timeoutrb.html

You might want to use channel.close or [session.shutdown!](https://www.rubydoc.info/gems/net-ssh/Net/SSH/Connection/Session#shutdown!-instance_method)

Reply:
I still want to keep the session (so that we wouldn't need to re-establish it in next iteration). I just want to exit from the Thread.

I made the following change;
Pulse thread sets ssh_thread_stop_flag, and if it set, we exit from the loop (and the thread) in the ssh thread.